### PR TITLE
Separate MatchOrder from SubgraphQuery

### PR DESCRIPTION
**Pull request by swreinehr**
Prepare for MatchOp Optimizer by moving some matching order information and associated functions from the subgraph query mining code into the matchop code so it can be passed in before matching.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -27,3 +27,6 @@ tags
 *.pyc
 galois.egg-info
 /_skbuild
+
+# ignore lonestar test files
+lonestar/querying/cpu
```
Comment (roshandathathri) [inline .gitignore:32]: Why is this required?

File: libquery/include/querying/ops/MatchOp.h [hunk 2]
```diff
@@ -51,6 +51,11 @@ class GALOIS_EXPORT MatchOp final : public Operator {
   //! Map from edge var name to the source and destination of that edge
   std::unordered_map<std::string, std::pair<uint32_t, uint32_t>>
       edge_var_to_src_dst_pair_;
+  //! vector storing order in which to query the query graph
+  std::vector<VertexId> matchingOrderToVertexMap;
+  //! reverse map of matchingOrderToVertexMap
+  std::vector<VertexId> vertexToMatchingOrderMap;
```
Comment (roshandathathri) [inline libquery/include/querying/ops/MatchOp.h:57]: Variable names use underscore instead of camelCase and member variables have trailing underscore, so matching_order_to_vertex_map_ and and vertex_to_matching_order_map_

File: libquery/src/SubgraphQuery.cpp [hunk 3]
```diff
@@ -40,16 +40,17 @@ class SubgraphQueryMiner {
...
   // const QueryEdgeProperties& query_edge_props_;
-  //! vector storing order in which to query the query graph
-  std::vector<VertexId> matchingOrderToVertexMap;
-  //! reverse map of matchingOrderToVertexMap
-  std::vector<VertexId> vertexToMatchingOrderMap;
+  //! vector storing order in which to query the query graph, pointer to
+  //! MatchOp's member
+  std::vector<VertexId>* matchingOrderToVertexMap;
+  //! reverse map of matchingOrderToVertexMap, pointer to MatchOp's member
+  std::vector<VertexId>* vertexToMatchingOrderMap;
```
Comment (roshandathathri) [inline libquery/src/SubgraphQuery.cpp:47]: Are these modified in this class? If they are not, then take const references.

File: libquery/include/querying/SubgraphQuery.h [hunk 4]
```diff
@@ -15,6 +15,7 @@ size_t SubgraphQuery(QueryGraph& query_graph,
                      const QueryEdgeProperties& query_edge_props,
                      const graphs::AttributedGraph& attributed_graph,
                      QueryEmbeddingList* returned_embeddings,
-                     std::vector<VertexId>* query_vertex_to_matching_order);
+                     std::vector<VertexId>* query_vertex_to_matching_order,
+                     std::vector<VertexId>* matching_order_to_query_vertex);
```
Comment (roshandathathri) [inline libquery/include/querying/SubgraphQuery.h:19]: Are these modified in this function? If they are not, then take const references and also change the order so that the return variable is at the end.

File: libquery/include/querying/OpGraph.h [hunk 5]
```diff
@@ -55,6 +55,7 @@ class GALOIS_EXPORT OpGraph {
 private:
   std::vector<std::unique_ptr<Operator>> ops_;
   std::unique_ptr<AggFuncMap> agg_func_map_;
+  friend class MatchOrderOptimizer;
```
Comment (roshandathathri) [inline libquery/include/querying/OpGraph.h:58]: Better to avoid friend class and have a clear interface.

File: libquery/include/querying/ops/MatchOp.h [hunk 6]
```diff
@@ -51,6 +51,11 @@ class GALOIS_EXPORT MatchOp final : public Operator {
...
+  friend class MatchOpBuilder;
```
Comment (roshandathathri) [inline libquery/include/querying/ops/MatchOp.h:58]: Why is this required?

**Review by roshandathathri — COMMENTED** [review]
Minor inline comments.

There are also seems to be some issue with new lines at the end of the file. Check if it is something to do with your git config autocrlf setting. If it is not, just add a newline at the end of the file.

**Comment by ddn0** [comment]
> There are also seems to be some issue with new lines at the end of the file. Check if it is something to do with your git config autocrlf setting. If it is not, just add a newline at the end of the file.

https://stackoverflow.com/questions/44704968/visual-studio-code-insert-new-line-at-the-end-of-files/44704969

**Comment by ddn0** [comment]
FYI, I have a bunch of changes to Subgraph.cpp over at #477

File: libquery/include/galois/graphs/AttributedGraph.h [hunk 7]
```diff
@@ -189,7 +189,8 @@ class GALOIS_EXPORT AttributedGraph {
   MatchQueryGraph(galois::QueryGraph& query_graph,
                   const galois::QueryNodeProperties& query_node_props,
                   const galois::QueryEdgeProperties& query_edge_props,
-                  std::vector<VertexId>* query_vertex_to_matching_order);
+                  std::vector<VertexId>* query_vertex_to_matching_order,
```
Comment (ddn0) [inline libquery/include/galois/graphs/AttributedGraph.h:192]: It might make sense to pull out these two vectors into an struct/class since they seem to move together, e.g.,
```cpp
struct MatchingOrder {
  std::vector<VertexId> query_vertex_to_order;
  std::vector<VertexId> order_to_query_vertex;
};
```

File: .gitignore (hunk 1 above)
Comment (swreinehr) [reply .gitignore:32]: That was an artifact of me having a weird git issue at 2 in the morning and just trying to merge things.

Basically, a head got detached and it was complaining about not having that folder committed. It should not be required, but it was something that was there on my branch that somehow got there and was never tracked. was there ever a file that went by that path?

Comment (ddn0) [reply .gitignore:32]: Yes there was but it was moved. No reason to change the gitignore though.

Comment (swreinehr) [reply .gitignore]: it was still in my directory, so I thought it was a remnant of something else. I've  removed it from the .gitignore

File: libquery/include/querying/SubgraphQuery.h [hunk 8]
```diff
@@ -2,18 +2,15 @@
...
-//!
-//! The matching order used is also returned to the caller for post processing.
 QueryEmbeddingList
 SubgraphQuery(QueryGraph& query_graph,
               const QueryNodeProperties& query_node_props,
               const QueryEdgeProperties& query_edge_props,
               const graphs::AttributedGraph& attributed_graph,
-              std::vector<VertexId>* query_vertex_to_matching_order);
-
+              const std::vector<VertexId>& query_vertex_to_matching_order,
```
Comment (ddn0) [inline libquery/include/querying/SubgraphQuery.h:14]: should this be a `MatchingOrder` as well?

File: libquery/include/querying/ops/MatchOp.h [hunk 9]
```diff
@@ -14,6 +14,22 @@
 
 namespace galois {
 
+//! Matching order for use in MatchOp, has some basic guarding
+//! keeps them both together aSDdasfasdfasdafds
```
Comment (ddn0) [inline libquery/include/querying/ops/MatchOp.h:18]: aSDdasfasdfasdafds is not an acronym I know about....

File: libquery/include/querying/ops/MatchOp.h [hunk 10]
```diff
@@ -14,6 +14,22 @@
...
+class MatchingOrder {
+private:
+  std::vector<VertexId> query_vertex_to_order_;
+  std::vector<VertexId> order_to_query_vertex_;
+
+public:
+  //! Set matching order from query to match, sets both, deletes input
+  void SetQueryToMatch(std::vector<VertexId>& query_to_match);
```
Comment (ddn0) [inline libquery/include/querying/ops/MatchOp.h:26]: I'd actually just leave this as a basic `struct` without any methods, 
```cpp
struct MatchingOrder
  std::vector<VertexId> query_vertex_to_order;
  std::vector<VertexId> order_to_query_vertex;
};
```

The accessor methods don't clarify the interface any.

File: libquery/include/querying/ops/MatchOp.h [hunk 11]
```diff
@@ -33,9 +49,9 @@ class GALOIS_EXPORT MatchOp final : public Operator {
   friend class MatchOpBuilder;
 
   std::vector<TableNodeName>
-  MakeNodeNames(const std::vector<uint32_t>& matching_order) const;
+  MakeNodeNames(const std::vector<galois::VertexId>& matching_order) const;
```
Comment (ddn0) [inline libquery/include/querying/ops/MatchOp.h:52]: take a MatchingOrder instead?

File: libquery/src/MatchOp.cpp [hunk 12]
```diff
@@ -81,6 +81,33 @@ MakeQueryNode(const galois::BitLabelSchema& schema, const std::string& label) {
...
 
+void galois::MatchingOrder::SetQueryToMatch(
+    std::vector<galois::VertexId>& query_to_match) {
+  auto n_vertices        = query_to_match.size();
+  query_vertex_to_order_ = std::move(query_to_match);
+  order_to_query_vertex_.resize(n_vertices);
+  for (auto i = n_vertices - 1; i; --i) {
+    order_to_query_vertex_[query_vertex_to_order_[i]] = i;
+  }
+}
```
Comment (ddn0) [inline libquery/src/MatchOp.cpp:92]: our autoformatter is nice but one thing it doesn't do is put newlines between functions. Unless you are defining a bunch of repeated, short functions, adding a newline usually helps future readability.

File: libquery/src/MatchOp.cpp [hunk 13]
```diff
@@ -167,10 +190,18 @@ std::unique_ptr<galois::Operator> galois::MatchOpBuilder::Build() {
...
       PrefixSumOutgoingEdges(num_query_nodes_, query_edges_);
 
+  //! this was removed at some point: ConstructVarNameToQueryNodeMap();
   if (!ConstructQueryGraph(prefix_sum)) {
     return MakeNoResult();
   }
-
+  //! initialize the matching order
+  InitializeMatchingOrder();
+  /*
```
Comment (ddn0) [inline libquery/src/MatchOp.cpp:199]: just remove this commented out code.

File: libquery/src/MatchOp.cpp [hunk 14]
```diff
@@ -167,10 +190,18 @@ std::unique_ptr<galois::Operator> galois::MatchOpBuilder::Build() {
   std::vector<size_t> prefix_sum =
       PrefixSumOutgoingEdges(num_query_nodes_, query_edges_);
 
+  //! this was removed at some point: ConstructVarNameToQueryNodeMap();
```
Comment (ddn0) [inline libquery/src/MatchOp.cpp:193]: - `//!`, `///` and `/**` all introduce doxygen comments that are intended to be put in the API docs. If you don't intend for this comment to be extracted (and it probably won't anyway because it is internal to a function), stick with `//` for comments.
- I'm not sure what this comment is supposed to convey. If this function is removed, should it be added? If so, why not add it now?

File: libquery/src/MatchOp.cpp [hunk 15]
```diff
@@ -167,10 +190,18 @@ std::unique_ptr<galois::Operator> galois::MatchOpBuilder::Build() {
...
   if (!ConstructQueryGraph(prefix_sum)) {
     return MakeNoResult();
   }
-
+  //! initialize the matching order
```
Comment (ddn0) [inline libquery/src/MatchOp.cpp:197]: Delete this comment. It doesn't say anymore than what the function is called.

File: libquery/src/MatchOp.cpp [hunk 16]
```diff
@@ -368,3 +399,79 @@ bool galois::MatchOpBuilder::ConstructQueryGraph(
 
   return true;
 }
+
+namespace {
+
+//! Ordering class: true if left vertex has higher degree than right vertex
+//! what this ends up doing is putting high degree vertices first
+//! used by InitializeMatchingOrder
+class OrderVertices {
```
Comment (ddn0) [inline libquery/src/MatchOp.cpp:408]: Consider renaming this class to better represent what it is doing: ordering by degree. E.g., OrderVerticesByDegree. Then you don't really need to put a doxygen comment on it at all, unless you want to give some context on why ordering by degree is useful.

Another note: because this class is in the unnamed namespace, readers can probably figure out where this class is used pretty easily, since it must be used in this translation unit. You might want to just omit "used by InitializeMatchingOrder" to save on future document maintenance.

File: libquery/src/MatchOp.cpp [hunk 17]
```diff
@@ -368,3 +399,79 @@ bool galois::MatchOpBuilder::ConstructQueryGraph(
...
+  }
+};
+} // namespace
+
+//! gets matching order of vertices in the query graph
+void galois::MatchOpBuilder::InitializeMatchingOrder() {
+  // assert(queryGraph.size() > 2);
+  // set_queryGraph.size()(queryGraph.size());
+  // set_num_patterns(1);
+  ///* no longer have to do this
```
Comment (ddn0) [inline libquery/src/MatchOp.cpp:425]: just delete unused code.

File: libquery/src/SubgraphQuery.cpp [hunk 18]
```diff
@@ -857,18 +795,19 @@ class SubgraphQueryMiner {
...
-galois::SubgraphQuery(QueryGraph& query_graph,
-                      const QueryNodeProperties& query_node_props,
-                      const QueryEdgeProperties& query_edge_props,
-                      const graphs::AttributedGraph& attributed_graph,
-                      std::vector<VertexId>* query_vertex_to_matching_order) {
+galois::QueryEmbeddingList galois::SubgraphQuery(
+    QueryGraph& query_graph, const QueryNodeProperties& query_node_props,
+    const QueryEdgeProperties& query_edge_props,
+    const graphs::AttributedGraph& attributed_graph,
+    const std::vector<VertexId>& query_vertex_to_matching_order,
```
Comment (ddn0) [inline libquery/src/SubgraphQuery.cpp:802]: `MatchingOrder` here?

File: libquery/src/MatchOp.cpp [hunk 19]
```diff
@@ -81,6 +81,33 @@ MakeQueryNode(const galois::BitLabelSchema& schema, const std::string& label) {
 
 } // namespace
 
+void galois::MatchingOrder::SetQueryToMatch(
+    std::vector<galois::VertexId>& query_to_match) {
+  auto n_vertices        = query_to_match.size();
+  query_vertex_to_order_ = std::move(query_to_match);
```
Comment (roshandathathri) [inline libquery/src/MatchOp.cpp:87]: Make the argument && to make it explicit to the user that the argument is moved.

File: libquery/src/MatchOp.cpp [hunk 20]
```diff
@@ -81,6 +81,33 @@ MakeQueryNode(const galois::BitLabelSchema& schema, const std::string& label) {
...
+  order_to_query_vertex_.resize(n_vertices);
+  for (auto i = n_vertices - 1; i; --i) {
+    order_to_query_vertex_[query_vertex_to_order_[i]] = i;
+  }
+}
+void galois::MatchingOrder::SetMatchToQuery(
+    std::vector<galois::VertexId>& match_to_query) {
+  auto n_vertices = match_to_query.size();
+  query_vertex_to_order_.resize(n_vertices);
+  order_to_query_vertex_ = std::move(match_to_query);
```
Comment (roshandathathri) [inline libquery/src/MatchOp.cpp:97]: Ditto

**Review by roshandathathri — COMMENTED** [review]
Minor nits

**Review by roshandathathri — APPROVED** [review]
LGTM

File: libquery/src/MatchingOrder.cpp [hunk 21]
```diff
@@ -0,0 +1,33 @@
+#include "querying/MatchingOrder.h"
+namespace galois {
```
Comment (ddn0) [inline libquery/src/MatchingOrder.cpp:2]: delete this namespace line.

It isn't needed when defining functions in cpp files (c.f., declaring them) and gets in the way of a benefit of using out-of-line definitions, detecting when a function is defined but not declared.

File: libquery/include/querying/ops/MatchOp.h (hunk 11 above)
Comment (swreinehr) [reply libquery/include/querying/ops/MatchOp.h:37]: The reason I haven't done this is because the one other use of it is in the dummyorder used for make empty, which doesn't really need to construct an inverse.

File: libquery/src/MatchingOrder.cpp (hunk 21 above)
Comment (swreinehr) [reply libquery/src/MatchingOrder.cpp]: done in 6d1b05a

File: libquery/src/MatchingOrder.cpp [hunk 22]
```diff
@@ -0,0 +1,30 @@
+#include "querying/MatchingOrder.h"
+void galois::MatchingOrder::SetQueryToMatch(
+    std::vector<galois::VertexId>&& query_to_match) {
+  auto n_vertices        = query_to_match.size();
+  query_vertex_to_order_ = std::move(query_to_match);
+  order_to_query_vertex_.resize(n_vertices);
+  for (auto i = n_vertices - 1; i; --i) {
```
Comment (ddn0) [inline libquery/src/MatchingOrder.cpp:7]: `size_t` (ditto and throughout).

`auto` is good to avoid writing a type multiple times, but if we use it to avoid writing a type at least once, it makes the code harder to read.

do we want to execute this loop if `i == 0`? and is it possible for `n_vertices == 0`? Perhaps `for .... i >= 0 ...`?

File: libquery/include/querying/ops/MatchOp.h [hunk 23]
```diff
@@ -53,6 +54,9 @@ class GALOIS_EXPORT MatchOp final : public Operator {
   //! Map from edge var name to the source and destination of that edge
   std::unordered_map<std::string, std::pair<uint32_t, uint32_t>>
       edge_var_to_src_dst_pair_;
+  //! vector storing order in which to query the query graph
+  MatchingOrder match_order_;
```
Comment (ddn0) [inline libquery/include/querying/ops/MatchOp.h:58]: nit: if we are going to call the variable `match_order_` maybe this class should be called `MatchOrder` instead of `MatchingOrder` (and rename throughout).

File: libquery/include/querying/ops/MatchOp.h (hunk 11 above)
Comment (ddn0) [reply libquery/include/querying/ops/MatchOp.h:37]: if the only concern is that we'd construct an inverse but never use it, I think the benefit of consistently using the abstraction of a `Match{ing}Order` to represent a match{ing} order would outweigh the cost since a query graph isn't expected to be millions of nodes and we don't expect to have more than a handful at a time.

If there are other reasons to not use this abstraction here, then feel free to leave things as is.

File: libquery/include/querying/ops/MatchOp.h (hunk 23 above)
Comment (swreinehr) [reply libquery/include/querying/ops/MatchOp.h]: it has been renamed.

**Review by ddn0 — APPROVED** [review]
:balloon: :flight_departure:
